File: apps/kernel/src/bin/chariox-cli/app_developer.rs
//! Early developer dispatch: no workspace lookup, Bun build, kernel or cwd change.

use std::{
    env,
    path::{Path, PathBuf},
    process::{Command, ExitCode, Stdio},
};

pub fn dispatch(args: &[String], protocol: u32) -> Option<Result<ExitCode, String>> {
    if args.first().map(String::as_str) != Some("app")
        || !args.get(1).is_some_and(|action| {
            matches!(
                action.as_str(),
                "create" | "keygen" | "manifest" | "pack" | "validate" | "inspect"
            )
        })
    {
        return None;
    }
    Some(run(&args[1..], protocol))
}

fn command_args(args: &[String], protocol: u32) -> Result<Vec<String>, String> {
    if args.len() > 64
        || args
            .iter()
            .any(|arg| arg.len() > 4096 || arg.contains('\0'))
    {
        return Err("App developer command arguments exceed their bounds".into());
    }
    if args
        .iter()
        .any(|arg| arg == "--kernel-protocol" || arg.starts_with("--kernel-protocol="))
    {
        return Err("The Chariox CLI supplies its current protocol; use chariox-app-package directly to select another contract.".into());
    }
    let mut forwarded = args.to_vec();
    if matches!(
        args.first().map(String::as_str),
        Some("create" | "manifest" | "pack" | "validate")
    ) && !args.iter().any(|arg| arg == "--help")
    {
        forwarded.extend(["--kernel-protocol".into(), protocol.to_string()]);
    }
    Ok(forwarded)
}

fn select_binary(explicit: Option<&Path>, launcher: &Path) -> Result<PathBuf, String> {
    if let Some(path) = explicit {
        if !path.is_absolute() || !path.is_file() {
            return Err("CHARIOX_APP_PACKAGE_BIN must be an absolute executable file path".into());
        }
        return Ok(path.to_owned());
    }
    let sibling = launcher
        .parent()
        .ok_or("cannot locate App developer tool")?
        .join("chariox-app-package");
    if sibling.is_file() {
        return Ok(sibling);
    }
    let installed = PathBuf::from("/usr/local/bin/chariox-app-package");
    if installed.is_file() {
        return Ok(installed);
    }
    Err("The App developer tool is not installed. Install the matching Chariox release, or build chariox-app-package and set CHARIOX_APP_PACKAGE_BIN to its absolute path.".into())
}

fn run(args: &[String], protocol: u32) -> Result<ExitCode, String> {
    let args = command_args(args, protocol)?;
    let explicit = env::var_os("CHARIOX_APP_PACKAGE_BIN")
        .filter(|value| !value.is_empty())
        .map(PathBuf::from);
    let executable = select_binary(
        explicit.as_deref(),
        &env::current_exe().map_err(|_| "cannot locate CLI executable")?,
    )?;
    let mut command = Command::new(executable);
    command
        .args(args)
        .stdin(Stdio::null())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit());
    // Keep original cwd so all developer-local relative paths mean what the
    // user typed. Replacing the launcher preserves status and signal behavior.
    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt;
        Err(format!(
            "App developer tool could not run: {}",
            command.exec()
        ))
    }
    #[cfg(not(unix))]
    {
        let status = command
            .status()
            .map_err(|error| format!("App developer tool could not run: {error}"))?;
        Ok(ExitCode::from(status.code().unwrap_or(1) as u8))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn arguments_keep_local_paths_literal_and_add_only_current_protocol() {
        let args = vec![
            "create".into(),
            "project ; $(touch forbidden)".into(),
            "--publisher".into(),
            "../public.json".into(),
        ];
        let mut expected = args.clone();
        expected.extend(["--kernel-protocol".into(), "500".into()]);
        assert_eq!(command_args(&args, 500).unwrap(), expected);
        assert!(command_args(
            &["pack".into(), "--kernel-protocol".into(), "1".into()],
            500
        )
        .is_err());
        assert_eq!(
            command_args(&["create".into(), "--help".into()], 500).unwrap(),
            ["create", "--help"]
        );
        assert!(dispatch(&["app".into(), "list".into()], 500).is_none());
    }
    #[test]
    fn explicit_relative_or_missing_helper_never_falls_back() {
        assert!(select_binary(
            Some(Path::new("chariox-app-package")),
            Path::new("/trusted/bin/chariox-cli")
        )
        .is_err());
        assert!(select_binary(
            Some(Path::new("/definitely-missing-chariox-app-package")),
            Path::new("/trusted/bin/chariox-cli")
        )
        .is_err());
    }
}

File: apps/cli/src/app-developer.ts
import { execFile } from "node:child_process"
import { constants } from "node:fs"
import { access, realpath, stat } from "node:fs/promises"
import { isAbsolute, join, resolve } from "node:path"
import { fileURLToPath } from "node:url"
import { LOCAL_DAEMON_PROTOCOL_VERSION } from "@chariox/kernel-client"

const commands = new Set(["create", "keygen", "manifest", "validate", "pack", "inspect"])
const protocolCommands = new Set(["create", "manifest", "validate", "pack"])
const maxOutputBytes = 2 * 1024 * 1024

export type AppDeveloperDeps = {
  locateBinary: () => Promise<string>
  execute: (binary: string, args: readonly string[]) => Promise<{ status: number, stdout: string }>
  write: (text: string) => void
}

export function isAppDeveloperCommand(action: string | undefined): boolean {
  return action !== undefined && commands.has(action)
}

export function defaultAppDeveloperDeps(write: (text: string) => void = (text) => { process.stdout.write(text) }): AppDeveloperDeps {
  return { locateBinary: () => locateAppPackageBinary(), execute: executePackageCommand, write }
}

/** No kernel connection, shell, dependency installation or implicit Cargo build. */
export async function runAppDeveloperCommand(
  args: readonly string[],
  deps: AppDeveloperDeps = defaultAppDeveloperDeps(),
): Promise<boolean> {
  const action = args[0]
  if (!isAppDeveloperCommand(action)) return false
  if (args.length > 64 || args.some((value) => Buffer.byteLength(value) > 4096 || value.includes("\0"))) {
    throw new Error("App developer command arguments exceed their bounds")
  }
  if (args.includes("--kernel-protocol") || args.some((value) => value.startsWith("--kernel-protocol="))) {
    throw new Error("The Chariox CLI supplies its current protocol; use chariox-app-package directly to select another contract.")
  }
  const command = [...args]
  if (protocolCommands.has(action!) && !args.includes("--help")) {
    command.push("--kernel-protocol", String(LOCAL_DAEMON_PROTOCOL_VERSION))
  }
  const response = await deps.execute(await deps.locateBinary(), command)
  if (Buffer.byteLength(response.stdout) > maxOutputBytes) throw new Error("App package tool exceeded its output limit")
  let value: unknown
  try { value = JSON.parse(response.stdout) } catch { throw new Error("App package tool returned an invalid response") }
  if (!value || typeof value !== "object" || !("ok" in value) || typeof value.ok !== "boolean") {
    throw new Error("App package tool returned an invalid response")
  }
  if (response.status !== 0 || !value.ok) {
    const error = "error" in value && value.error && typeof value.error === "object" ? value.error : undefined
    if (error && "code" in error && typeof error.code === "string" && "message" in error && typeof error.message === "string") {
      throw new Error(`${error.code}: ${error.message}`)
    }
    throw new Error("App package command failed")
  }
  if (!("result" in value)) throw new Error("App package tool returned an invalid response")
  deps.write(`${JSON.stringify(value.result, null, 2)}\n`)
  return true
}

export type PackedAppPackage = { appId: string, version: string, packageDigest: string }

/**
 * `app pack` with the CLI's own protocol, returning its result instead of printing it.
 * The helper verifies the archive with the installer's verifier before writing it.
 */
export async function packAppPackage(
  options: { bundle: string, manifest: string, key: string, output: string },
  deps: Omit<AppDeveloperDeps, "write"> = defaultAppDeveloperDeps(),
): Promise<PackedAppPackage> {
  let printed = ""
  await runAppDeveloperCommand(
    ["pack", "--bundle", options.bundle, "--manifest", options.manifest, "--key", options.key, "--output", options.output],
    { ...deps, write: (text) => { printed += text } },
  )
  const result = JSON.parse(printed) as { manifest?: { appId?: unknown, version?: unknown }, packageDigest?: unknown }
  const { appId, version } = result.manifest ?? {}
  if (typeof appId !== "string" || typeof version !== "string" || typeof result.packageDigest !== "string" || !/^sha256:[0-9a-f]{64}$/.test(result.packageDigest)) {
    throw new Error("App package tool returned an invalid pack result")
  }
  return { appId, version, packageDigest: result.packageDigest }
}

async function executable(path: string): Promise<string | undefined> {
  try {
    const selected = await realpath(path)
    if (!(await stat(selected)).isFile()) return undefined
    await access(selected, constants.X_OK)
    return selected
  } catch { return undefined }
}

export async function locateAppPackageBinary(
  env: NodeJS.ProcessEnv = process.env,
  moduleUrl: string = import.meta.url,
): Promise<string> {
  const explicit = env.CHARIOX_APP_PACKAGE_BIN?.trim()
  if (explicit) {
    if (!isAbsolute(explicit)) throw new Error("CHARIOX_APP_PACKAGE_BIN must be an absolute executable path")
    const selected = await executable(explicit)
    if (!selected) throw new Error("CHARIOX_APP_PACKAGE_BIN is not an executable file")
    return selected
  }
  // A source checkout is derived from this module, never the user's cwd or PATH.
  const repository = fileURLToPath(new URL("../../../", moduleUrl))
  const checkout = await stat(join(repository, "packages/app-package/Cargo.toml")).then((entry) => entry.isFile(), () => false)
  const fromTarget = async (target: string) => {
    for (const profile of ["release", "debug"]) {
      const selected = await executable(join(target, profile, "chariox-app-package"))
      if (selected) return selected
    }
    return undefined
  }
  if (checkout && env.CARGO_TARGET_DIR?.trim()) {
    const selected = await fromTarget(resolve(repository, env.CARGO_TARGET_DIR.trim()))
    if (selected) return selected
  }
  const installed = await executable("/usr/local/bin/chariox-app-package")
  if (installed) return installed
  if (checkout) {
    const selected = await fromTarget(join(repository, "target"))
    if (selected) return selected
  }
  throw new Error("The App developer tool is not installed. Install the matching Chariox release, or build chariox-app-package in the source checkout and set CHARIOX_APP_PACKAGE_BIN to its absolute path.")
}

function executePackageCommand(binary: string, args: readonly string[]): Promise<{ status: number, stdout: string }> {
  return new Promise((resolve, reject) => {
    execFile(binary, [...args], { encoding: "utf8", maxBuffer: maxOutputBytes, timeout: 120_000 }, (error, stdout) => {
      if (error && typeof error.code !== "number") {
        reject(new Error("App package tool could not run or exceeded its output/time limit"))
      } else {
        resolve({ status: error ? Number(error.code) : 0, stdout })
      }
    })
  })
}
